serial: uart_native_pty: IRQ support

tpambor opened the pull request:
Add support for the interrupt-driven API. Interrupts are emulated using a polling thread.

This was merged in a previous version as #93957 but caused problems, see #94425 and reverted in #94426.

The issue stemmed from false RX interrupts being triggered when `select` indicated that data was available to read, but the subsequent read operation failed. In the updated implementation, a dedicated polling thread now feeds incoming data into a ring buffer. Interrupts are no longer triggered directly by select; instead, the ring buffer is used to signal valid RX interrupts, ensuring more reliable and accurate handling.

(Locally) this is now passing `samples/sensor/sensor_shell` and `tests/drivers/can/host`, which previously failed in #94425.

tpambor committed:
serial: uart_native_pty: IRQ support

Add support for the interrupt-driven API. Interrupts are
emulated using a polling thread.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>
Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

dcpleung commented:
I wonder why CI didn't catch it last time.

tpambor commented:
> I wonder why CI didn't catch it last time.

`tests/drivers/can/host` isn't run in CI.

For `samples/sensor/sensor_shell`, there was also 7c0e4aecace9287bcd0508e991672a732fc95ddc, which helped with the symptoms and was sufficient to pass. With, this PR sensor shell would actually pass with 7c0e4aecace9287bcd0508e991672a732fc95ddc reverted but I think it doesn't do no bad, so there is no reason to revert.

henrikbrixandersen reviewed (changes requested):
This change makes tests using the native_sim UART with pytest and the twister_harness Shell much slower than before.

As an example, running `tests/drivers/can/host` without this change:

`INFO    - 1/1 native_sim/native/64      drivers.can.host                                   PASSED (native 1.865s <host>)`

But with this change:

`INFO    - 1/1 native_sim/native/64      drivers.can.host                                   PASSED (native 40.506s <host>)`

tpambor commented:
> This change makes tests using the native_sim UART with pytest and the twister_harness Shell much slower than before.
> 
> As an example, running `tests/drivers/can/host` without this change:
> 
> `INFO - 1/1 native_sim/native/64 drivers.can.host PASSED (native 1.865s <host>)`
> 
> But with this change:
> 
> `INFO - 1/1 native_sim/native/64 drivers.can.host PASSED (native 40.506s <host>)`

This performance degradation is due to a locking issue in the shell code. Enabling `CONFIG_LOG=y` and by default `CONFIG_SHELL_LOG_BACKEND=y` which changes how locking in shell works, results in performance comparable to the polling implementation.

aescolar reviewed (changes requested):
I think this approach is too complicated and heavy (i.e. no need for ring buffers).
I think instead something that would work:
When we want to check if there is something ready in stdin, we read 1 character.
If we read() == 0, we know we reached EOF. We flag it and never try to read again.
If we read() == -1, we know there was nothing there yet.
if we read() == 1, we save that character we just read ahead in data (`data->char_read_ahead` or whatever), and set a flag (`data->stdin_ready_char` or whatever).
Next time we want to check if there is data we first check this flag (if the flag is set we already answer yes, otherwise we do the check with one character above).

When there is an attempt to actually read data by the user (len > 0).
If that flag (`data->stdin_ready_char`) was set we first copy that character to the user buffer, clear `data->stdin_ready_char`, and continue trying to read `len-1`.
if the flag was not set, we attempt to read len. If we get read() == 0 => we flag eof.

That should solve the issue, and reduce a lot the number of syscalls.
We should probably check if we need to do the select() at all. Or if we can just use the read() to detect if data is available in stdin or not. (for a real pty we don't need the select).

clamattia commented:
> I think this approach is too complicated and heavy (i.e. no need for ring buffers). I think instead something that would work: When we want to check if there is something ready in stdin, we read 1 character. If we read() == 0, we know we reached EOF. We flag it and never try to read again. If we read() == -1, we know there was nothing there yet. if we read() == 1, we save that character we just read ahead in data (`data->char_read_ahead` or whatever), and set a flag (`data->stdin_ready_char` or whatever). Next time we want to check if there is data we first check this flag (if the flag is set we already answer yes, otherwise we do the check with one character above).
> 
> When there is an attempt to actually read data by the user (len > 0). If that flag (`data->stdin_ready_char`) was set we first copy that character to the user buffer, clear `data->stdin_ready_char`, and continue trying to read `len-1`. if the flag was not set, we attempt to read len. If we get read() == 0 => we flag eof.
> 
> That should solve the issue, and reduce a lot the number of syscalls. We should probably check if we need to do the select() at all. Or if we can just use the read() to detect if data is available in stdin or not. (for a real pty we don't need the select).

How does removing the ring-buffer help. Would we not expect better performance with the ring-buffer (for example less system calls?) How would your approach help?

The performance issue may be unrelated to the driver.

Some thoughts:

1. Maybe someone could make an analysis using [tracing](https://docs.zephyrproject.org/latest/services/tracing/index.html)
2. `can_shell` is storing the `const struct shell *` and using it later from the work-queue to implement "async"-shell commands. This seems a bit unique to me, not sure if that way of doing "async"-shell commands is supported/recommended. An attempt to make them synchronous was rejected.
3. There is probably a deadlock (or live lock) situation in the `can_shell`, that can be worked around by `CONFIG_SHELL_TX_TIMEOUT_MS=0` but needs actual fixing due degrading performance with that approach, when `CONFIG_LOG` is disabled. Without `CONFIG_SHELL_TX_TIMEOUT_MS=0` we have the deadlock with `CONFIG_LOG` enabled.

henrikbrixandersen commented:
The updated patch greatly improves the performance, but it is still significantly slower than when using polling (which seems odd from a user perspective):

As an example, running `tests/drivers/can/host` without this change:

`INFO    - 1/1 native_sim/native/64      drivers.can.host                                   PASSED (native 1.865s <host>)`

But with this updated patch:

`INFO    - 1/1 native_sim/native/64      drivers.can.host                                   PASSED (native 3.797s <host>)`

tpambor commented:
> The updated patch greatly improves the performance, but it is still significantly slower than when using polling (which seems odd from a user perspective):
> 
> As an example, running `tests/drivers/can/host` without this change:
> 
> `INFO - 1/1 native_sim/native/64 drivers.can.host PASSED (native 1.865s <host>)`
> 
> But with this updated patch:
> 
> `INFO - 1/1 native_sim/native/64 drivers.can.host PASSED (native 3.797s <host>)`

This behavior isn’t too surprising, under the hood, the IRQ emulation also relies on polling. However, it introduces an additional abstraction layer, which leads to significantly more context switches. Data is transferred in blocks of up to 8 bytes between the shell task and the IRQ emulation task, which then polls the data out.

That said, this PR is not focused on performance improvements. Its goal is to enhance feature completeness, enabling testing of device drivers or other functionality that only supports IRQ mode.

If shell performance is a concern for this sample, you can set SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN=n. Alternatively, we could consider making this the default for native-sim, so there would be no performance regressions by default, and users could opt in to use IRQ mode when needed.

henrikbrixandersen commented:
> That said, this PR is not focused on performance improvements. Its goal is to enhance feature completeness, enabling testing of device drivers or other functionality that only supports IRQ mode.

Sure, but if introducing a new feature negatively impacts performance of an existing feature, we cannot just pull it in and say we'll address the introduced performance issue at a later time.

> If shell performance is a concern for this sample, one can set `SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN=n`. Alternatively, we could consider making this the default for native-sim, so there would be no performance regressions by default, and users could opt in to use IRQ mode when needed.

I reckon most developers will be using the native_sim UART for logging and shell functionality, so defaulting to polling there might not be a bad idea. I'll leave that up to @aescolar.

aescolar commented:
@tpambor thanks. I was out yesterday, I will take a look at this next week and probably ping you in discord to ask about the motivation for some of the choices.

aescolar commented:
> So the approach would be to merge your developements of the irq-mode @aescolar but make it default to polling mode on native_sim so as not to impact performance of most users, that do not want to explicitly test a driver that requires irq-mode?

I thinks so yes.

tpambor committed:
shell: backend: Default to poll mode for uart_native_pty

The interrupt-driven UART API is emulated via polling on native_sim,
which introduces additional overhead. Defaulting to poll mode improves
performance by avoiding this emulation cost.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

dcpleung commented:
> @dcpleung @jakub-uC given how the shell interacts with this driver in irq mode, I wonder if there is something misunderstood about the driver API in either side..

Come to think of it, maybe using UART ASYNC API would be better for this then the pure interrupt API?
